### ISSUE TITLE
unnecessary_to_owned includes invalid `&` in suggestion

### ISSUE THREAD

**Issue by tmpolaczyk** [rust-lang/rust-clippy#9771]
### Summary

The `unnecessary_to_owned` lint includes an invalid `&` in the suggestion, making the code not compile. Removing that reference makes the suggestion valid.

### Reproducer

I tried this code:

```rust
use std::marker::PhantomData;

pub struct Key<K: AsRef<[u8]>, V: ?Sized>(K, PhantomData<V>);

impl<K: AsRef<[u8]>, V: ?Sized> Key<K, V> {
    pub fn new(key: K) -> Key<K, V> {
        Key(key, PhantomData)
    }
}

pub fn pkh(pkh: &[u8]) -> Key<Vec<u8>, String> {
    Key::new([b"pkh-", pkh].concat().to_vec())
}
```

I expected to see this happen:

The lint should say that `.to_vec()` is redundant and can be removed.

Instead, this happened:

```
warning: unnecessary use of `to_vec`
  --> src/main.rs:13:14
   |
13 |     Key::new([b"pkh-", pkh].concat().to_vec())
   |              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: use: `&[b"pkh-", pkh].concat()`
   |
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#unnecessary_to_owned
   = note: `#[warn(clippy::unnecessary_to_owned)]` on by default
```

It is almost correct, but the suggestion includes a `&`, which is invalid because `Key::new` expects a `Vec<u8>`, not a `&Vec<u8>`. Removing that leading `&` makes the suggestion valid, and the code compiles.

### Version

```text
rustc 1.67.0-nightly (11ebe6512 2022-11-01)
binary: rustc
commit-hash: 11ebe6512b4c77633c59f8dcdd421df3b79d1a9f
commit-date: 2022-11-01
host: x86_64-unknown-linux-gnu
release: 1.67.0-nightly
LLVM version: 15.0.4
```

**Comment by koka831** [comment]
I reproduced it with bit reduced example:

```rust
pub struct Key<K: AsRef<[u8]>>(K);

pub fn from(c: &[u8]) -> Key<Vec<u8>> {
    let v = [c].concat();
    Key(v.to_vec())
}
```

with output:

```terminal
warning: unnecessary use of `to_vec`
 --> src/main.rs:5:9
  |
5 |     Key(v.to_vec())
  |         ^^^^^^^^^^ help: use: `&v`
  |
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#unnecessary_to_owned
  = note: `#[warn(clippy::unnecessary_to_owned)]` on by default
```
